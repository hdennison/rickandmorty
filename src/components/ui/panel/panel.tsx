import { Slot } from "@radix-ui/react-slot";
import { cva, type VariantProps } from "class-variance-authority";
import { cn } from "../utils";

const panelVariants = cva("rounded-lg", {
  variants: {
    padding: {
      none: "p-0",
      medium: "p-4",
      large: "p-8",
    },
    level: {
      default: "bg-[var(--color-surface)] shadow-md",
      low: "bg-[var(--color-surface-low)]",
    },
  },
  defaultVariants: {
    padding: "medium",
    level: "default",
  },
});

export interface PanelProps
  extends React.HTMLAttributes<HTMLDivElement>,
    VariantProps<typeof panelVariants> {
  asChild?: boolean;
}

export const Panel = ({
  className,
  children,
  level,
  padding,
  asChild = false,
  ...props
}: PanelProps) => {
  const Comp = asChild ? Slot : "div";

  return (
    <Comp
      className={cn(panelVariants({ className, level, padding }))}
      {...props}
    >
      {children}
    </Comp>
  );
};
