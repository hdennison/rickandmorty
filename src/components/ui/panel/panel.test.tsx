import { expect, test } from "vitest";
import { render } from "vitest-browser-react";
import { Panel } from "./panel";

test("Renders as wrapper", () => {
  const panel = render(<Panel id="panel">Foo</Panel>);

  // Snapshots are not the best solution, it woould be much better to set up Storybook + Chromatic
  expect(panel).toMatchSnapshot();
});

test("Renders asChild", () => {
  const panel = render(
    <Panel id="Panel" asChild>
      <article>Bar</article>
    </Panel>,
  );

  // Snapshots are not the best solution, it woould be much better to set up Storybook + Chromatic
  expect(panel).toMatchSnapshot();
});
