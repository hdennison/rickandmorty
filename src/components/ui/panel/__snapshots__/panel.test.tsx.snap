// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renders as wrapper 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    
    
    
  


    <div>
      <div
        class="bg-[var(--color-surface)] rounded-lg shadow-md p-4"
        id="panel"
      >
        Foo
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bg-[var(--color-surface)] rounded-lg shadow-md p-4"
      id="panel"
    >
      Foo
    </div>
  </div>,
  "debug": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholder": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Renders asChild 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    
    
    
  


    <div>
      <article
        class="bg-[var(--color-surface)] rounded-lg shadow-md p-4"
        id="Panel"
      >
        Bar
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="bg-[var(--color-surface)] rounded-lg shadow-md p-4"
      id="Panel"
    >
      Bar
    </article>
  </div>,
  "debug": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholder": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
